offset.go

```diff
@@ -0,0 +1,414 @@
+
+	if config.offset.args.set == "" {
+		config.offset.set = -1
+	} else if config.offset.args.set == "newest" {
+		config.offset.setNewest = true
+	} else if config.offset.args.set == "oldest" {
+		config.offset.setOldest = true
+	} else if newOffset, err := strconv.Atoi(config.offset.args.set); err == nil {
+		config.offset.set = int64(newOffset)
+	} else {
+		fmt.Fprintf(os.Stderr, "Invalid value for setting the offset. possible values are \"oldest\", \"newest\", or any numerical value. err=%s\n", err)
+		os.Exit(2)
+	}
+
+	config.offset.verbose = config.offset.args.verbose
+	config.offset.version = kafkaVersion(config.offset.args.version)
+}
+
+func offsetRun(closer chan struct{}) {
+	var err error
+	if config.offset.verbose {
+		sarama.Logger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+
+	conf := sarama.NewConfig()
+	conf.Version = config.offset.version
+	u, err := user.Current()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to read current user err=%v", err)
+	}
+	conf.ClientID = "kt-offset-" + u.Username
+	if config.offset.verbose {
+		fmt.Fprintf(os.Stderr, "sarama client configuration %#v\n", conf)
+	}
+
+	client, err := sarama.NewClient(config.offset.brokers, conf)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create client err=%v\n", err)
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	om, err := sarama.NewOffsetManagerFromClient(config.offset.group, client)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create offset manager for [%s] err=%v\n", config.offset.group, err)
+		os.Exit(1)
+	}
+
+	getBroker := client.Coordinator
+	getPartitions := client.Partitions
+	getOffset := client.GetOffset
+	getPartitionOffsetManager := om.ManagePartition
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	out := make(chan string)
+	go func(out chan string) {
+		topics, err := client.Topics()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to read topics err=%v\n", err)
+			os.Exit(1)
+		}
+
+		for _, t := range topics {
+			if config.offset.args.topic == "" || len(config.offset.topic.FindString(t)) > 0 {
+				offsetsForTopic(getBroker, getPartitions, getOffset, getPartitionOffsetManager, config.offset.group, t, out)
+			}
+		}
+
+		wg.Done()
+	}(out)
+
+	// print to console
+	go func() {
+		for {
+			select {
+			case m := <-out:
+				fmt.Println(m)
+			}
+		}
+	}()
+	wg.Wait()
+}
+
+// offsetsForTopic processes the offsets for a given topic
+func offsetsForTopic(
+	getBroker func(string) (*sarama.Broker, error),
+	getPartitions func(string) ([]int32, error),
+	getOffset func(string, int32, int64) (int64, error),
+	getPartitionOffsetManager func(string, int32) (sarama.PartitionOffsetManager, error),
+	group string,
+	topic string,
+	out chan string,
+) {
+	partitions, err := getPartitions(topic)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to read partitions for [%s] err=%v\n", topic, err)
+		os.Exit(1)
+	}
+
+	for _, p := range partitions {
+		if config.offset.partition == -1 || config.offset.partition == p {
```

Comment:
That is correct but the partition flag is a filter, so `-partition=<non-existing partition number>` returns 0 results.

It works the same for topics, when adding `-topic=something-non-existing`, then it will return 0 results